contracts/TokenHolder.sol

```diff
@@ -441,18 +597,107 @@ contract TokenHolder is MultiSigWallet {
+     * @param _v V of the signature.
+     * @param _r R of the signature.
+     * @param _s S of the signature.
+     *
+     * @return ephemeralKey_ which is bool execution status of
+     *         coGateway.redeem.
      */
-    function bytesToBytes4(bytes _input) public pure returns (bytes4 out_) {
-        require(
-            _input.length >= 4,
-            "Input bytes length is less than 4."
+    function verifyRedeemExecutableTransaction(
+        uint256 _amount,
+        address _beneficiary,
+        uint256 _gasPrice,
+        uint256 _gasLimit,
+        uint256 _redeemerNonce,
+        address _coGateway,
+        uint256 _nonce,
+        uint8 _v,
+        bytes32 _r,
+        bytes32 _s
+    )
+        private
+        returns (address ephemeralKey_)
+    {
+        bytes memory redeemData = abi.encodeWithSelector(
+            REDEEM_SIGNED_DATA_CALLPREFIX,
+            _amount,
+            _beneficiary,
+            _gasPrice,
+            _gasLimit,
+            _redeemerNonce
         );
 
-        for (uint8 i = 0; i < 4; i++) {
-            out_ |= bytes4(_input[i] & 0xFF) >> (i * 8);
-        }
+        (, ephemeralKey_) = verifyExecutableTransaction(
+            REDEEM_RULE_CALLPREFIX,
+            _coGateway,
+            redeemData,
+            _nonce,
+            _v,
+            _r,
+            _s
+        );
     }
+
+    /**
+     * @notice Executes CoGateway redeem after approving amount to CoGateway.
+     *
+     * @param _coGateway CoGateway contract address.
+     * @param _amount Redeem amount that will be transferred from tokenholder
+     *                account.
+     * @param _beneficiary The address in the origin chain where the tokens
+     *                     will be released.
+     * @param _gasPrice Gas price that tokenholder is ready to pay to get the
+     *                  redemption process done.
+     * @param _gasLimit Gas limit that tokenholder is ready to pay.
+     * @param _redeemerNonce Nonce of the redeemer address.
+     * @param _hashLock Hash Lock provided by the facilitator.
+     *
+     * @return executionStatus_ which is bool execution status of
+     *         coGateway.redeem.
+     */
+    function executeRedeem(
+        address _coGateway,
+        uint256 _amount,
+        address _beneficiary,
+        uint256 _gasPrice,
+        uint256 _gasLimit,
+        uint256 _redeemerNonce,
+        bytes32 _hashLock
+    )
+        private
+        returns (bool executionStatus_)
+    {
+        token.approve(_coGateway, _amount);
+
+        bytes memory data = abi.encodeWithSelector(
+            COGATEWAY_REDEEM_CALLPREFIX,
+            _amount,
+            _beneficiary,
+            msg.sender,
+            _gasPrice,
+            _gasLimit,
+            _redeemerNonce,
+            _hashLock
+        );
+
+        // solium-disable-next-line security/no-call-value
+        executionStatus_ = _coGateway.call.value(msg.value)(data);
```

Comment:
We had discussion around it. Reason for using .call is to consume Ekey nonce in case CoGateway.redeem fails.
Here is the concluded discussion comment by Amulya in the current PR itself:

https://github.com/OpenSTFoundation/openst-contracts/pull/84#issuecomment-436216594

Reply:
As discussed separately, I believe you will add clarifying comments inline here.

But, for the avoidance of doubt, the documented conclusion of an internal discussion (in a location that is subject to revision, such as a comment to an issue) is not sufficient to make clear to a user the sensibility of a design choice that is, on its face, peculiar.

Reply:
Updated above comment for more clarity.